src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,386 @@
+    CompactionPhase,
+    InvalidDatasetKindError,
+    NullCompactionListener,
+};
+use futures::stream::TryStreamExt;
+use kamu_core::compact_service::CompactService;
+use kamu_core::*;
+use opendatafabric::{
+    DatasetHandle,
+    DatasetKind,
+    DatasetName,
+    MetadataEvent,
+    Multihash,
+    OffsetInterval,
+};
+use url::Url;
+
+use crate::*;
+
+pub struct CompactServiceImpl {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    dataset_authorizer: Arc<dyn domain::auth::DatasetActionAuthorizer>,
+    run_info_dir: PathBuf,
+}
+
+#[derive(Debug, Clone)]
+struct DataSliceBatchInfo {
+    pub data_slices_batch: Vec<Url>,
+    pub prev_offset: Option<u64>,
+    pub new_offset_interval: Option<OffsetInterval>,
+    pub new_file_path: Option<PathBuf>,
+}
+
+struct ChainFilesInfo {
+    block_file_urls: Vec<Url>,
+    new_head: Multihash,
+    data_slice_batches: Vec<DataSliceBatchInfo>,
+}
+
+#[component(pub)]
+#[interface(dyn CompactService)]
+impl CompactServiceImpl {
+    pub fn new(
+        dataset_authorizer: Arc<dyn domain::auth::DatasetActionAuthorizer>,
+        dataset_repo: Arc<dyn DatasetRepository>,
+        run_info_dir: PathBuf,
+    ) -> Self {
+        Self {
+            dataset_repo,
+            dataset_authorizer,
+            run_info_dir,
+        }
+    }
+
+    async fn get_chain_files_info(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        dataset_name: &DatasetName,
+        compact_dir_path: &Path,
+        max_slice_size: u64,
+        max_slice_records: u64,
+    ) -> Result<ChainFilesInfo, CompactError> {
+        // Declare mut values for result
+
+        let mut new_head: Option<Multihash> = None;
+        let mut block_file_urls: Vec<Url> = vec![];
+        let mut data_slice_batch_info: DataSliceBatchInfo = DataSliceBatchInfo {
+            prev_offset: None,
+            new_offset_interval: None,
+            data_slices_batch: vec![],
+            new_file_path: None,
+        };
+        let mut data_slice_batches: Vec<DataSliceBatchInfo> = vec![];
+        let (mut batch_size, mut new_end_offset_interval, mut batch_records): (u64, u64, u64) =
+            (0, 0, 0);
+
+        ////////////////////////////////////////////////////////////////////////////////
+
+        let chain = dataset.as_metadata_chain();
+        let head = chain.resolve_ref(&BlockRef::Head).await?;
+        let mut block_stream = chain.iter_blocks_interval(&head, None, false);
+        let object_data_repo = dataset.as_data_repo();
+        let object_block_repo = chain.as_metadata_block_repository();
+
+        while let Some((block_hash, block)) = block_stream.try_next().await? {
+            block_file_urls.push(object_block_repo.get_internal_url(&block_hash).await);
+
+            match block.event {
+                MetadataEvent::AddData(add_data_event) => {
+                    if let Some(output_slice) = &add_data_event.new_data {
+                        new_head = block.prev_block_hash.clone();
```

Comment:
I don't fully understand the role of this `new_head`, perhaps, the name does not match to it's purpose.
Note that the blocks are iterated in reverse direction - from the newest to oldes, so it maybe is an "old_head".
Looking at the algorithm, it will point to the last node before the very first `AddData`.

Reply:
Yeah,  you are right this is exactly the last node before the very first `AddData`. Sorry for confusing name, In my mind it will be a new head after resetting old blocks. But `old_head` sounds better